File: packages/sdks/server-sdk/src/types/events.ts
/**
 * Event Types for BullMQ
 */

// ============================================================================
// Standard Event Types
// ============================================================================

export type StandardEventType =
  | "gameworld.created"
  | "gameworld.updated"
  | "gameworld.deleted"
  | "gameworld.order_changed"
  | "popup.created"
  | "popup.updated"
  | "popup.deleted"
  | "survey.created"
  | "survey.updated"
  | "survey.deleted"
  | "survey.settings.updated"
  | "maintenance.settings.updated"
  | "maintenance.started"
  | "maintenance.ended"
  | "whitelist.updated"
  | "client_version.created"
  | "client_version.updated"
  | "client_version.deleted"
  | "banner.created"
  | "banner.updated"
  | "banner.deleted"
  | "service_notice.created"
  | "service_notice.updated"
  | "service_notice.deleted"
  | "store_product.created"
  | "store_product.updated"
  | "store_product.deleted"
  | "store_product.bulk_updated"
  | "environment.created"
  | "environment.deleted"
  | "feature_flag.changed"
  | "feature_flag.created"
  | "feature_flag.updated"
  | "feature_flag.deleted"
  | "segment.created"
  | "segment.updated"
  | "segment.deleted";

export interface StandardEventData {
  id?: number | string; // Optional for bulk events that don't have a single id
  timestamp: number;
  /**
   * Environment identifier (environmentName value).
   * This is the standard external identifier for environments.
   */
  environment?: string;
  isVisible?: boolean | number; // For gameworld.updated, popup.updated events (MySQL returns 0/1)
  isActive?: boolean | number; // For survey.updated, store_product.updated events (MySQL returns 0/1)
  status?: string; // For banner.created, banner.updated events (draft, published, archived)
  isMaintenance?: boolean; // For maintenance.settings.updated events
  maintenanceStartDate?: string; // For maintenance.settings.updated events
  maintenanceEndDate?: string; // For maintenance.settings.updated events
  maintenanceMessage?: string; // For maintenance.settings.updated events
  maintenanceLocales?: Array<{ lang: string; message: string }>; // For maintenance.settings.updated events
  count?: number; // For bulk events (store_product.bulk_updated)
  // Full object data for direct cache updates (avoiding list refresh)
  clientVersion?: any; // For client_version.created, client_version.updated events
  serviceNotice?: any; // For service_notice.created, service_notice.updated events
  segmentName?: string; // For segment.created, segment.updated, segment.deleted events
}

export interface StandardEvent {
  type: StandardEventType;
  data: StandardEventData;
}

// ============================================================================
// Custom Event Types
// ============================================================================

export interface CustomEvent {
  type: string; // Custom event type (e.g., 'custom:player.levelup')
  data: any; // Custom data
  timestamp: number;
}

// ============================================================================
// SDK Event Types (for SDK event listeners)
// ============================================================================

export interface SdkEvent {
  type: string;
  data: any;
  timestamp: string; // ISO8601 format
}

// ============================================================================
// Event Listener Types
// ============================================================================

export type EventCallback = (event: SdkEvent) => void | Promise<void>;

export interface EventListenerMap {
  [eventType: string]: EventCallback[];
}

File: packages/sdks/server-sdks/gatrix-go-server-sdk/types/events.go
package types

// StandardEventType is a known SDK event type
type StandardEventType string

const (
	EventGameWorldCreated      StandardEventType = "game_world.created"
	EventGameWorldUpdated      StandardEventType = "game_world.updated"
	EventGameWorldDeleted      StandardEventType = "game_world.deleted"
	EventGameWorldOrderChanged StandardEventType = "game_world.order_changed"
	EventPopupCreated          StandardEventType = "popup.created"
	EventPopupUpdated          StandardEventType = "popup.updated"
	EventPopupDeleted          StandardEventType = "popup.deleted"
	EventSurveyCreated         StandardEventType = "survey.created"
	EventSurveyUpdated         StandardEventType = "survey.updated"
	EventSurveyDeleted         StandardEventType = "survey.deleted"
	EventSurveySettingsUpdated StandardEventType = "survey.settings.updated"
	EventMaintenanceUpdated    StandardEventType = "maintenance.settings.updated"
	EventWhitelistUpdated      StandardEventType = "whitelist.updated"
	EventStoreProductCreated   StandardEventType = "store_product.created"
	EventStoreProductUpdated   StandardEventType = "store_product.updated"
	EventStoreProductDeleted   StandardEventType = "store_product.deleted"
	EventStoreProductBulk      StandardEventType = "store_product.bulk_updated"
	EventFeatureFlagChanged    StandardEventType = "feature_flag.changed"
	EventFeatureFlagCreated    StandardEventType = "feature_flag.created"
	EventFeatureFlagUpdated    StandardEventType = "feature_flag.updated"
	EventFeatureFlagDeleted    StandardEventType = "feature_flag.deleted"
	EventSegmentCreated        StandardEventType = "segment.created"
	EventSegmentUpdated        StandardEventType = "segment.updated"
	EventSegmentDeleted        StandardEventType = "segment.deleted"
)

// StandardEventData holds data for standard events
type StandardEventData struct {
	ID                   interface{}         `json:"id,omitempty"`
	Timestamp            int64               `json:"timestamp"`
	EnvironmentID        string              `json:"environmentId,omitempty"`
	IsVisible            interface{}         `json:"isVisible,omitempty"` // bool or 0/1
	IsActive             interface{}         `json:"isActive,omitempty"`  // bool or 0/1
	Status               string              `json:"status,omitempty"`
	IsMaintenance        *bool               `json:"isMaintenance,omitempty"`
	Count                *int                `json:"count,omitempty"`
	SegmentName          string              `json:"segmentName,omitempty"`
	Segment              interface{}         `json:"segment,omitempty"`
	ProjectID            string              `json:"projectId,omitempty"`
	ChangedKeys          []string            `json:"changedKeys,omitempty"`
	ChangeType           string              `json:"changeType,omitempty"`
	MaintenanceStartDate string              `json:"maintenanceStartDate,omitempty"`
	MaintenanceEndDate   string              `json:"maintenanceEndDate,omitempty"`
	MaintenanceMessage   string              `json:"maintenanceMessage,omitempty"`
	MaintenanceLocales   []MaintenanceLocale `json:"maintenanceLocales,omitempty"`
}

// StandardEvent is a Redis Pub/Sub event
type StandardEvent struct {
	Type string            `json:"type"`
	Data StandardEventData `json:"data"`
}

// SdkEvent is the event passed to user callbacks
type SdkEvent struct {
	Type      string      `json:"type"`
	Data      interface{} `json:"data"`
	Timestamp string      `json:"timestamp"`
}

// EventCallback is a function that handles SDK events
type EventCallback func(event SdkEvent)

// ToBool converts the isVisible/isActive interface value to bool
func ToBool(v interface{}) bool {
	if v == nil {
		return false
	}
	switch val := v.(type) {
	case bool:
		return val
	case float64:
		return val != 0
	case int:
		return val != 0
	default:
		return false
	}
}
